ui/src/containers/SampleFlexView.jsx

```diff
@@ -1,156 +1,245 @@
-import React from 'react';
+import { useEffect, useState } from 'react';
 import { Col } from 'react-bootstrap';
-import { connect } from 'react-redux';
+import { useDispatch, useSelector } from 'react-redux';
 
 import { filterAction } from '../actions/sampleGrid';
 
-class NewSampleFlexView extends React.Component {
-  constructor(props) {
-    super(props);
+export default function SampleFlexView(props) {
+  const {
+    displayPuckCellContextMenu,
+    cellMenuID,
+    puckMenuID,
+    type = 'FLEX HCD',
+  } = props;
 
-    this.renderCircle = this.renderCircle.bind(this);
-    this.isCellSelected = this.isCellSelected.bind(this);
+  const filterOptions = useSelector((state) => state.sampleGrid.filterOptions);
+  const scContent = useSelector((state) => state.sampleChanger.contents);
+  const [pendingEvent, setPendingEvent] = useState(null);
+
+  const numberOfCell = scContent.children.length; // Number of cells/slices
+
+  const dispatch = useDispatch();
+
+  function handleClickonCellPuck(event, cellID, puckID) {
+    dispatch(
+      filterAction({ cellFilter: `${cellID}`, puckFilter: `${puckID}` }),
+    );
+    event.stopPropagation();
   }
 
-  componentDidMount() {
-    document.addEventListener('keydown', this.onKeyDown, false);
+  function isCellSelected(cellID) {
+    return Number(filterOptions.cellFilter) === cellID;
   }
 
-  onClickCell(event, idx, disabled) {
-    if (!disabled) {
-      this.props.filter({ cellFilter: `${idx + 1}` });
+  function isPuckSelected(cellID, puckID) {
+    return (
+      Number(filterOptions.cellFilter) === cellID &&
+      Number(filterOptions.puckFilter) === puckID
+    );
+  }
+
+  // Call displayPuckCellContextMenu only after Redux state updates
+  useEffect(() => {
+    if (pendingEvent) {
+      displayPuckCellContextMenu(
+        pendingEvent.event,
+        pendingEvent.menuID,
+        pendingEvent.cellID,
+        pendingEvent.puckID,
+      );
+      setPendingEvent(null);
+    }
+  }, [filterOptions, pendingEvent, displayPuckCellContextMenu]);
+
+  useEffect(() => {
+    if (filterOptions.cellFilter === '') {
+      dispatch(filterAction({ cellFilter: `${1}`, puckFilter: '' }));
     }
+  }, [filterOptions.cellFilter, dispatch]);
+
+  function handleDisplayPuckCellContextMenu(event, menuID, cellID, puckID) {
+    event.preventDefault();
+    handleClickonCellPuck(event, cellID, puckID === null ? '' : puckID);
```

Comment:
Could you not just dispatch the action directly here (and `await` it as suggested in my previous comment)?

Reply:
I think i try an async await and it did not work.  could not see why.

Reply:
If all else fails, `setTimeout(() => { displayPuckCellContextMenu(...) }, 0)` after `await dispath(...)` might work as a slightly cleaner workaround.

Reply:
very Good suggestion, i try await again and it work without the the need of setTimeout.
When i try it the first time it behave differently maybe because of something else at that time. 

but i was happy to know the hack by the way AHAHa